### ISSUE TITLE
watchOS Complication

### ISSUE THREAD

**Issue by durosity** [g-mocken/ZoeStatus#1]
Kudos on a wonderfully made app!  It beats the pants off the official one!

On watchOS 7 Zoe Status now shows up in the complications list on the settings app, but I've not found any watch faces where I can actually add that option.  Is this supposed to work or is it a bug/undeveloped feature?  I'd love to be able to see my Zoe charge level % and/or range in the corners of the  infograph watch face!

**Comment by g-mocken** [comment]
That is indeed an "undeveloped feature", i.e. it is basically implemented, but I could not get periodic updates to work properly, i.e. the charge level in the complication would be basically stuck and only act as a short cut into the app. 
Periodic updates do works in the simulator, but on a real watch, probably due to Apple's power saving policies, my code is never called to refresh the value shown in the complication.
That's why the feature is currently disabled until someone shows me how to make it work.

**Comment by coolio2004** [comment]
I use complications on watch faces only for starting the app. For me there is no reason to show any status.

**Comment by g-mocken** [comment]
I have submitted a new version to TestFlight.

**Comment by durosity** [comment]
Oooh exciting!  What version is it?

**Comment by g-mocken** [comment]
v1.4.0.0:
"The iOS app has only received under-the-hood changes (network code) and should work exactly as before.
The watchOS app, although mostly looking the same, was re-written using SwiftUI. The long press for accessing the menu is therefore replaced by an explicit button that links to a simple menu. The change is necessary, because Apple appears to be dropping watchOS support for the old style menus.
As an experimental feature, there are now a couple of watch complications. Do not expect frequent updates of the complications: although one update every 15 minutes is requested, the operating system will often not grant it that many."

**Comment by g-mocken** [comment]
This is what they are called in code:

`supportedFamilies: [ .utilitarianSmall, .utilitarianSmallFlat, .utilitarianLarge, .modularSmall, .modularLarge]
`

I am not sure, how exactly that translates to watch face styles. Here's an excerpt from Apple's documentation:

_.utilitarianSmall
A small square or rectangular area that ClockKit Displays on the Utility, Mickey, Chronograph, and Simple watch faces.
.utilitarianSmallFlat
A small rectangular area that ClockKit Displays on the Photos, Motion, and Timelapse watch faces.
.utilitarianLarge
A large rectangular area that spans the width of the screen in the Utility and Mickey watch faces.
.modularSmall
A small square area that ClockKit displays on the Modular watch face.
.modularLarge
A large rectangular area that ClockKit displays on the Modular watch face._

(If they are working, it is not difficult to add a few more.)

**Comment by durosity** [comment]
Interesting, not showing for me on Modular.  Also I’m noticing I can’t actually refresh the info on the app itself, no option appears on screen when I press/hold.

**Comment by durosity** [comment]
I’m a moron.  I even read that earlier and just completely forgot about it. 😅

However, I’m having an issue.  I go in, select the option, it comes up with the expected options.  I select the option to “Request new credentials” -> open iOS app -> “Go” -> “Success”… then.. nothing happens.  If I attempt to refresh I get the message stating “no user credentials present”.

**Comment by g-mocken** [comment]
But you did enter the credentials in the iOS app (settings) before trying to transfer them to the watch, didn’t you?

**Comment by durosity** [comment]
Yes, the iOS app is fully configured and working.  I have also deleted the iOS app (and watch) and reinstalled from TestFlight and reconfigured, confirmed the iOS side is working and then tried the watchOS app again but same behaviour.

Now I am on watchOS 10.6.1, rather than 11.. is it possible that’s related?  Both to the issue with the data and also the widgets missing?

**Comment by g-mocken** [comment]
Thanks for trying it out and you feedback.

Regarding the credentials, I think I have located the issue – unfortunately it was working correctly on any device that was previously used with an earlier version ... but in a freshly installed simulator, I can reproduce it. I can probably  upload a fix for that tomorrow.

Regarding the complications, it turns out that I apparently missed a deprecation warning:

```
getComplicationDescriptors(handler:)
Returns the list of complication descriptors.
watchOS 7.0–9.0
Deprecated

```
That's probably why it works on my series 3 (watchOS 8). In a watchOS 11 simulator I also could not select any complications. To work around this will probably take a longer.

**Comment by g-mocken** [comment]
New TestFlight v1.4.0.1 available.

**Comment by g-mocken** [comment]
The next TestFlight version v1.4.1.0 will have new-style complications for watchOS 9 and above. However, as I currently do not own a more recent watch, I could only test them in the simulator. I do not know if or how often they are updated on a real device.
I have just submitted it, so it should appear in TestFlight soon...

**Comment by dbhorst** [comment]
Hey, the TestFlight version is expired and it looks like the bug is not fixed in the App Store version yet. 
Can you renew the TestFlight or release the version to the App Store?

**Comment by derdast** [comment]
> Hey, the TestFlight version is expired and it looks like the bug is not fixed in the App Store version yet. Can you renew the TestFlight or release the version to the App Store?

Regarding the iOS app: The new version is available on the App Store and is working properly. However, data became corrupted during the update process. I had to uninstall and reinstall the app. It works well now.

**Comment by dbhorst** [comment]
To be more specific: The iOS widget and watch complication does not work, always showing 0%. Credentials are transferred to the watch and the iOS and watchOS app shows the correct data.
Tried reinstalling (from AppStore) again and had no luck. I found the version number in the Settings dialog. Indeed it is the same as in the TestFlight (Store only shows 1.4.5 and not 1.4.5.3), my bad.

**Comment by g-mocken** [comment]
Yes, AppStore (1.4.5) provides exactly the same version (1.4.5.3) as latest TestFlight, it just does not show the last version string digit.

The watch complication shows 0% when it cannot read the real data from MyR. This occasionally happens to me (whenever MyR backend is not working properly), too, but it usually updates correctly again after a couple of hours.

When you tap the complication to open the app, then refresh there, do you see a more verbose error in the app? 
If not, does the complication copy the app's status when you switch back to the watch home screen afterwards? Because it should.

**Comment by dbhorst** [comment]
Tried all of that. Watch app works perfectly. After days it fixed itself today. Apple wants me to buy a new watch. Thanks for the feedback!
